### PR TITLE
0.0.1

### PR DESCRIPTION
**Pull request by kfirfer**
Draft pull request for comparison

### CODE REVIEW

File: charts/topolvm/templates/certificates/certificates.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
+# Generate a CA Certificate used to sign certificates for the webhook
+apiVersion: cert-manager.io/v1alpha3
+kind: Certificate
+metadata:
+  name: webhook-ca
```
Comment (kfox1111) [inline charts/topolvm/templates/certificates/certificates.yaml]: while less important for namespaced resources, probably all resource names should be templated with the release name.

Lots of places in the PR that needs it.

File: charts/topolvm/templates/controller/deployment.yaml [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            - "--feature-gates=Topology=true"
+            - --enable-leader-election
+            - --leader-election-namespace={{ .Release.Namespace }}
+            - --leader-election-type=leases
+          volumeMounts:
+            - name: socket-dir
+              mountPath: /run/topolvm
+
+        - name: csi-attacher
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
```
Comment (kfox1111) [inline charts/topolvm/templates/controller/deployment.yaml:72]: These images should probably be separate entries, so the upstream proxies can be used. They use different image names from each other.

File: charts/topolvm/templates/scheduler/psp.yaml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  hostNetwork: true
+  runAsUser:
+    rule: 'MustRunAsNonRoot'
+  seLinux:
+    rule: 'RunAsAny'
+  supplementalGroups:
+    rule: 'MustRunAs'
+    ranges:
+      - min: 1
+        max: 65535
```
Comment (kfox1111) [inline charts/topolvm/templates/scheduler/psp.yaml:36]: Can this be tightened up as we know what uid/gid's the scheduler is running under?

File: charts/topolvm/templates/storageclass.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+{{ if .Values.storageclass.create }}
```
Comment (kfox1111) [inline charts/topolvm/templates/storageclass.yaml]: maybe consider wrapping this in a {{range $key, $valule := .Values.storageClasses }}  and make it a hash with the name being the key? Then multiple storage classes could be defined with the chart

File: charts/topolvm/values.yaml [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+  nodeSelector: {}
+
+  ## terminationGracePeriodSeconds
+  ##
+  terminationGracePeriodSeconds: 10
+
+  ## Define resources requests and limits for single Pods.
+  ## ref: https://kubernetes.io/docs/user-guide/compute-resources/
+  resources:
+    requests:
```
Comment (kfox1111) [inline charts/topolvm/values.yaml]: Its best practice to leave out the actual values but instead comment them out for reference.

File: charts/topolvm/values.yaml [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+# CSI controller service
+controller:
+  replicaCount: 2
+  ## terminationGracePeriodSeconds
+  ##
+  terminationGracePeriodSeconds: 10
+  ## Define resources requests and limits for single Pods.
+  ## ref: https://kubernetes.io/docs/user-guide/compute-resources/
+  resources:
+    requests:
```
Comment (kfox1111) [inline charts/topolvm/values.yaml]: same

File: charts/topolvm/values.yaml [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+  ## Ref: https://kubernetes.io/docs/user-guide/node-selection/
+  ##
+  nodeSelector: {}
+
+
+
+# Whether to create storageclass
+# ref: https://kubernetes.io/docs/concepts/storage/storage-classes/
+storageclass:
+  create: true
```
Comment (kfox1111) [inline charts/topolvm/values.yaml]: see comment above about multiple storage classes

File: charts/topolvm/values.yaml [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+  additionalParameters: {}
+#    "topolvm.cybozu.com/device-class": "ssd"
+
+
+
+# Container Security Context
+# ref: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/
+securityContext:
+  runAsUser:  10000
+  runAsGroup: 10000
```
Comment (kfox1111) [inline charts/topolvm/values.yaml:171]: Maybe these values can be used directly in the psp to tighten it up

File: charts/topolvm/values.yaml [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+# Enable pod security policy
+# ref: https://kubernetes.io/docs/concepts/policy/pod-security-policy/
+podSecurityPolicy:
+  create: true
+
+
+crd:
+  # Weather to create CRD
+  create: true
```
Comment (kfox1111) [inline charts/topolvm/values.yaml]: In my experience helm has been very inconsistent on its behavior regarding safe handling of crds between versions (even patch versions). I personally don't consider it safe to ever use helm to manage a crd and I don't recommend it to others. Should this be false by default?

**Review by kfox1111 — COMMENTED** [review]
A great start for a chart! Thanks for working on this. A few comments inline.

File: charts/topolvm/templates/scheduler/psp.yaml (hunk 3 above)
Comment (kfirfer) [reply charts/topolvm/templates/scheduler/psp.yaml:36]: I think the container havnt specific user
@ymmt2005  Is it right  ?

```
$ cat /etc/passwd
root:x:0:0:root:/root:/bin/bash
daemon:x:1:1:daemon:/usr/sbin:/usr/sbin/nologin
bin:x:2:2:bin:/bin:/usr/sbin/nologin
sys:x:3:3:sys:/dev:/usr/sbin/nologin
sync:x:4:65534:sync:/bin:/bin/sync
games:x:5:60:games:/usr/games:/usr/sbin/nologin
man:x:6:12:man:/var/cache/man:/usr/sbin/nologin
lp:x:7:7:lp:/var/spool/lpd:/usr/sbin/nologin
mail:x:8:8:mail:/var/mail:/usr/sbin/nologin
news:x:9:9:news:/var/spool/news:/usr/sbin/nologin
uucp:x:10:10:uucp:/var/spool/uucp:/usr/sbin/nologin
proxy:x:13:13:proxy:/bin:/usr/sbin/nologin
www-data:x:33:33:www-data:/var/www:/usr/sbin/nologin
backup:x:34:34:backup:/var/backups:/usr/sbin/nologin
list:x:38:38:Mailing List Manager:/var/list:/usr/sbin/nologin
irc:x:39:39:ircd:/var/run/ircd:/usr/sbin/nologin
gnats:x:41:41:Gnats Bug-Reporting System (admin):/var/lib/gnats:/usr/sbin/nologin
nobody:x:65534:65534:nobody:/nonexistent:/usr/sbin/nologin
_apt:x:100:65534::/nonexistent:/usr/sbin/nologin
```

File: charts/topolvm/values.yaml (hunk 8 above)
Comment (kfirfer) [reply charts/topolvm/values.yaml:171]: I think leaving the user to decide will be more generic, no ?

File: charts/topolvm/values.yaml (hunk 9 above)
Comment (kfirfer) [reply charts/topolvm/values.yaml]: True
Set to false as default

File: charts/topolvm/templates/controller/deployment.yaml (hunk 2 above)
Comment (kfirfer) [reply charts/topolvm/templates/controller/deployment.yaml:72]: I'm sorry, I didn't fully understand

File: charts/topolvm/templates/scheduler/psp.yaml (hunk 3 above)
Comment (ymmt2005) [reply charts/topolvm/templates/scheduler/psp.yaml:36]: In `passwd`, yes.  For PSP, users and groups must be specified as numeric IDs.

Comment (kfirfer) [reply charts/topolvm/templates/scheduler/psp.yaml:36]: I mean, There isnt a dedicated user in the container for topolvm , right ?

Comment (ymmt2005) [reply charts/topolvm/templates/scheduler/psp.yaml:36]: Right.

File: charts/topolvm/values.yaml (hunk 8 above)
Comment (kfox1111) [reply charts/topolvm/values.yaml:171]: yeah. I mean, use {{ .Values.securityContext.runAsUser }} and group in the psp to tighten up the uid / groups allowed to run as. Then whatever the user specifies here is applied to the psp.

**Pull request closed without merging**